Python/thread.c

```diff
@@ -71,6 +72,98 @@ PyThread_init_thread(void)
 #endif
 
 
+/*
+ * Lock support.
+ */
+
+PyThread_type_lock
+PyThread_allocate_lock(void)
+{
+    if (!initialized) {
+        PyThread_init_thread();
+    }
+
+    PyMutex *lock = (PyMutex *)PyMem_RawMalloc(sizeof(PyMutex));
+    if (lock) {
+        *lock = (PyMutex){0};
+    }
+
+    return (PyThread_type_lock)lock;
+}
+
+void
+PyThread_free_lock(PyThread_type_lock lock)
+{
+    PyMem_RawFree(lock);
+}
+
+PyLockStatus
+PyThread_acquire_lock_timed(PyThread_type_lock lock, PY_TIMEOUT_T microseconds,
+                            int intr_flag)
+{
+    PyTime_t timeout;  // relative timeout
+    if (microseconds >= 0) {
+        // bpo-41710: PyThread_acquire_lock_timed() cannot report timeout
+        // overflow to the caller, so clamp the timeout to
+        // [PyTime_MIN, PyTime_MAX].
+        //
+        // PyTime_MAX nanoseconds is around 292.3 years.
+        //
+        // _thread.Lock.acquire() and _thread.RLock.acquire() raise an
+        // OverflowError if microseconds is greater than PY_TIMEOUT_MAX.
+        timeout = _PyTime_FromMicrosecondsClamp(microseconds);
+    }
+    else {
+        timeout = -1;
+    }
+
+    _PyLockFlags flags = _Py_LOCK_DONT_DETACH;
+    if (intr_flag) {
+        flags |= _PY_LOCK_HANDLE_SIGNALS;
+    }
```

Comment:
I don't think `intr_flag` maps directly to `_PY_LOCK_HANDLE_SIGNALS`.

`_PY_LOCK_HANDLE_SIGNALS` invokes `Py_MakePendingCalls()` on interrupt and then returns `PY_LOCK_INTR` on error.

`intr_flag` just returned `PY_LOCK_INTR` on interrupt

I think we need a new flag for this behavior.

Reply:
> _PY_LOCK_HANDLE_SIGNALS invokes Py_MakePendingCalls() on interrupt and then returns PY_LOCK_INTR on error. intr_flag just returned PY_LOCK_INTR on interrupt. I think we need a new flag for this behavior.

Oh, I didn't notice this difference. I added _PY_FAIL_IF_INTERRUPTED flag for that.